### PR TITLE
Disable registration flows on SSO servers

### PR DESCRIPTION
**Pull request by turt2live**
https://github.com/matrix-org/matrix-react-sdk/pull/4169 to develop. Note that the change is slightly different due to SSO changes in the Login page on develop.

### CODE REVIEW

File: src/components/structures/auth/Login.js [hunk 1]
```diff
@@ -151,6 +151,18 @@ export default createReactClass({
         return this.state.busy || this.props.busy;
     },
 
+    _isSsoStep: function() {
+        return this._getCurrentFlowStep() === 'm.login.sso' || this._getCurrentFlowStep() === 'm.login.cas';
+    },
+
+    _getSsoUrl: function(kind) {
+        if (kind === 'm.login.cas') {
+            return this._loginLogic.getSsoLoginUrl("cas");
```
Comment (t3chguy) [inline src/components/structures/auth/Login.js]: this method no longer exists in latest develop.

File: src/components/structures/auth/Login.js [hunk 2]
```diff
@@ -339,6 +351,19 @@ export default createReactClass({
         this.props.onRegisterClick();
     },
 
+    onTryRegisterClick: function(ev) {
+        if (this._isSsoStep()) {
+            // If we're showing SSO it means that registration is also probably disabled,
+            // so intercept the click and instead pretend the user clicked 'Sign in with SSO'.
+            ev.preventDefault();
+            ev.stopPropagation();
+            window.location = this._getSsoUrl(this._getCurrentFlowStep());
```
Comment (t3chguy) [inline src/components/structures/auth/Login.js]: this should just be a call to `platform.startSingleSignOn(...)`

File: src/components/structures/auth/Login.js [hunk 3]
```diff
@@ -339,6 +344,20 @@ export default createReactClass({
         this.props.onRegisterClick();
     },
 
+    onTryRegisterClick: function(ev) {
+        if (this._isSsoStep()) {
```
Comment (t3chguy) [inline src/components/structures/auth/Login.js]: there's something I don't like about calling `_getCurrentFlowStep` three times for one click of a button

Comment (turt2live) [reply src/components/structures/auth/Login.js]: Yea, it made more sense when we were using it in a couple places on the release branch. I've internalized it now.

**Pull request merged**